scrapy/core/downloader/handlers/__init__.py

```diff
@@ -20,25 +26,45 @@ def __init__(self, crawler):
             # component (extension, middleware, etc).
             if clspath is None:
                 continue
-            cls = load_object(clspath)
-            try:
-                dh = cls(crawler.settings)
-            except NotConfigured as ex:
-                self._notconfigured[scheme] = str(ex)
-            else:
-                self._handlers[scheme] = dh
+            self._schemes[scheme] = clspath
 
         crawler.signals.connect(self._close, signals.engine_stopped)
 
+    def _get_handler(self, scheme):
+        """Lazy-load the downloadhandler for a scheme
+        only on the first request for that scheme.
+        """
+        if scheme in self._handlers:
+            return self._handlers[scheme]
+        if scheme in self._notconfigured:
+            return None
+        if scheme not in self._schemes:
+            self._notconfigured[scheme] = 'no handler available for that scheme'
+            return None
+
+        path = self._schemes[scheme]
+        try:
+            dhcls = load_object(path)
+            dh = dhcls(self._crawler_settings)
+        except NotConfigured as ex:
+            self._notconfigured[scheme] = str(ex)
+            return None
+        except Exception as ex:
+            logger.exception('Loading "{}" for scheme "{}" handler'
+                             .format(path, scheme))
```

Comment:
Shouldn't we pass crawler or spider here? I think it is also better to pass path and scheme as parameters, without calling `format` ourselves.